src/core/exec-invoke.c

```diff
@@ -2234,6 +2234,37 @@ static int setup_private_users(uid_t ouid, gid_t ogid, uid_t uid, gid_t gid) {
         return 0;
 }
 
+static int setup_private_pids(const ExecContext *c, ExecParameters *p) {
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r, q;
+
+        if (unshare(CLONE_NEWPID) < 0)
+                return log_exec_debug_errno(c, p, errno, "Failed to unshare pid namespace: %m");
+
+        /* The first process created after unsharing a pid namespace becomes PID 1 in the pid namespace, so
+         * we have to fork after unsharing the pid namespace to become PID 1. The parent sends the child
+         * pidref to the manager and exits while the child process continues with the rest of exec_invoke()
+         * and finally executes the actual payload. */
+
+        r = pidref_safe_fork("(sd-pidfd-child)", /*flags=*/ 0, &pidref);
+        if (r < 0)
+                return log_exec_debug_errno(c, p, r, "Failed to fork child into new pid namespace: %m");
+
+        if (r > 0) {
+                /* In the parent process, we send the child pidref to the manager and exit. */
+
+                if (p->pidref_transport_fd >= 0) {
+                        q = send_one_fd_iov(p->pidref_transport_fd, pidref.fd, &IOVEC_MAKE(&pidref.pid, sizeof(pidref.pid)), 1, 0);
+                        if (q < 0)
+                                return log_exec_debug_errno(c, p, q, "Failed to send child pidref to manager: %m");
+                }
+        }
```

Comment:
you are not actually exiting here?

Reply:
ah, you are doing that in the parent. please add a comment about this. `/* Our caller will make sure the "middle" process will exit here quickly. */`